librosa/core/pitch.py

```diff
@@ -336,3 +339,278 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+    y : np.ndarray [shape=(n,)]
+        audio time series.
+
+    sr : number > 0 [scalar]
+        sampling rate of `y` in Hertz.
+
+    win_length : int [scalar]
+        length of the window in samples.
+        By default, win_length=1024 corresponds to a time scale of 46 ms at
+        a sampling rate of 22050 Hz.
+
+    hop_length : None or int > 0 [scalar]
+        number of audio samples between adjacent pYIN predictions.
+        If `None`, defaults to `frame_length // 4`.
+
+    fmin: number > 0 [scalar]
+        minimum frequency in Hertz.
+
+    fmax: number > 0 [scalar]
+        maximum frequency in Hertz.
+
+    n_thresholds : int [scalar]
+        number of thresholds for peak estimation.
+
+    n_bins_per_semitone : int [scalar]
+        number of bins per semitone for discretizing the pitch space
+        between fmin and fmax.
+
+    Returns
+    -------
+    f0: np.ndarray [shape=(n_frames,)]
+        time series of fundamental frequencies in Hertz.
+
+    voiced_prob: np.ndarray [shape=(n_frames,)]
+        time series containing the probability that a frame is voiced.
+
+
+    Examples
+    --------
+    Computing a fundamental frequency curve from an audio input:
+    >>> y = librosa.chirp(440, 880, duration=7.0)
+    >>> f0, voiced_prob = pyin(y)
+    '''
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = int(win_length // 4)
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Autocorrelation.
+    y_frames = util.frame(y, frame_length=2*win_length, hop_length=hop_length)
+    n_frames = y_frames.shape[-1]
+    a = np.fft.fft(y_frames, 2*win_length, axis=0)
+    b = np.fft.fft(y_frames[win_length::-1, :], 2*win_length, axis=0)
+    acf_frames = np.fft.ifft(a * b, 2*win_length, axis=0).real[win_length:]
+
+    # Energy terms.
+    window = scipy.signal.windows.boxcar(win_length + 1)
+    energy_frames = scipy.signal.fftconvolve(
+        y_frames**2, window[:, None], mode="valid", axes=0)
+
+    # Difference function.
+    yin_frames = energy_frames[0, :] + energy_frames - 2*acf_frames
+
+    # Cumulative mean normalized difference function.
+    min_period = np.maximum(int(np.floor(sr / fmax)), 2)
+    max_period = np.minimum(int(np.ceil(sr / fmin)), win_length - 1)
+    yin_numerator = yin_frames[min_period - 1:max_period + 1, :]
+    tau_range = np.arange(1, max_period + 1)[:, None]
+    cumulative_mean = np.cumsum(yin_frames[1:max_period + 1, :], axis=0) / tau_range
+    yin_denominator = cumulative_mean[min_period - 2:max_period, :]
+    yin_frames = yin_numerator / (yin_denominator + util.tiny(yin_denominator))
+
+    # Parabolic interpolation.
+    parabola_a = (yin_frames[:-2, :] + yin_frames[2:, :] - 2*yin_frames[1:-1, :]) / 2
+    parabola_b = (yin_frames[2:, :] - yin_frames[:-2, :]) / 2
+    parabolic_shifts = -parabola_b / (2*parabola_a + util.tiny(parabola_a))
+    parabolic_shifts[np.abs(parabolic_shifts) > 1] = 0
```

Comment:
is there a way that we could write a share `__yin` subroutine that gets called by both `yin` and `pyin` ?